### ISSUE TITLE
mp4box generates uneven representations when dashing

### ISSUE THREAD

**Issue by sandersaares** [gpac/gpac#521]
When provided input data of different duration (audio/video/subtitles), mp4box generates a DASH presentation with missing data. For example, using the sample data referenced below, the result is:
- 38 audio segments (152 nominal seconds of data)
- 43 audio segments (172 nominal seconds of data)
- 10 subtitle segments (40 nominal seconds of data)

With a period duration of 151.84 marked in the manifest.

~~While this behavior is OK in terms of video/audio - I do expect it to take the shortest duration, as that can fill all the representations with data and be conforming to DASH and, soon, CMAF - mp4box does not currently cover subtitles with acceptable behavior.~~

_Edit: behavior is also uneven with audio/video. None of the lengths are synchronized as expected. Note that the source material has uneven GOP sizes, which makes proper segmenting perhaps hard, so SegmentTemplate timing does not work out well. Still, the issue is demonstrated sufficiently by this test data._

Indeed, the same behavior would even not be appropriate to extend - it is normal not to have subtitles until the end! You have credits etc near the end, with no text samples. This is normal and extending the same behavior seen above would lead to undesired cutting.

Instead, my expectation is that mp4box generates empty subtitle segments to fill the subtitle track to the end. This is fairly trivial with WebVTT, at least, by including a single "empty cue" sample. With TTML, possibly it is also trivial? Though I wonder how it interacts with the different TTML-derived formats (EBU-TT versus IMSC1 and such).

mp4box version: gpac-0.6.2-DEV-rev517-gd4d3966-master-x64
Uploaded zip file with sample input data as "GitHub report: mp4box generates uneven representations" (uneven.zip).
Command to reproduce: mp4box -dash 4000 -profile "dashavc264:live" -out out\Manifest.mpd audio.mp4 video.mp4 subtitles.mp4

**Comment by sandersaares** [comment]
[uneven.zip](https://github.com/gpac/gpac/files/274041/uneven.zip)

**Comment by rbouqueau** [comment]
Hi,

> While this behavior is OK in terms of video/audio - I do expect it to take the shortest duration, as that can fill all the representations with data and be conforming to DASH and, soon, CMAF - mp4box does not currently cover subtitles with acceptable behavior.
> 
> Indeed, the same behavior would even not be appropriate to extend - it is normal not to have subtitles until the end! You have credits etc near the end, with no text samples. This is normal and extending the same behavior seen above would lead to undesired cutting.

It seems to me like it's what MP4Box does already. The period duration is the A/V one and there are less segments for subtitles (that can indeed stop before).

> my expectation is that mp4box generates empty subtitle segments

I think we agree on the fact that mp4box won't create empty subtitle samples. That would be a modification of semantics (like adding black frames or silent audio) and MP4Box is not the right tool for that.

You can force the duration of the subtitle input with ```:dur=```:
```MP4Box -dash 10000  -profile dashavc264:live audio.mp4 video.mp4 subtitles.mp4:dur=152```

Would it suit your need?

Also what problem are you trying to solve? Is there a compatibility issue with players?

**Comment by sandersaares** [comment]
Yes, most players encounter serious problems when there are 404s in a DASH presentation. The overall assumption is that a DASH presentation never has missing data in its timeline.

The implementations I have seen vary in behavior but playback often stops at the first 404 so, in effect, the sample data provided in this issue would stop at around 40 seconds. Some players have even stopped earlier, when loading the 404 into the buffer! Some only fail when seeking into the "404" timespan but work OK on linear playback into it. And so forth.

The issue as I see it is that current mp4box behavior inconsistent right now with regard to A/V and subtitles. As a result, it produces DASH presentations that often fail to play.

I disagree that generating empty samples is inappropriate - though I admit it is unreasonably complex for audio/video. For subtitles, it seems a good fit and is, I suspect, also what most users would expect.

The bottom line is that there does not appear to be an obvious way to ensure that a DASH presentation generated by mp4box does not have any 404s. Manually pre-processing the data streams to handle timing/alignment seems excessively difficult, as it also has to consider things like dash segment boundaries and potentially a lot of low-level timing details. After all, isn't correct segmenting and timing the main function mp4box actually serves in the dashing process?

Manually forcing a duration onto the streams seems like a bit of a hack, although if it works then it can (with some troublesome data pre-processing) be used, I suppose. However, I note that when I do specify a subtitle duration, it simply duplicates the last segment forever. This seems wrong - the subtitle segments at the end should be empty.

**Comment by rbouqueau** [comment]
> The issue as I see it is that current mp4box behavior inconsistent right now with regard to A/V and subtitles. As a result, it produces DASH presentations that often fail to play.

I don't understand what is "inconsistent". The period duration is the longest duration of any media. What did I miss?

In my opinion that's the MP4Box user responsibility to prepare its content, whether by checking GOP alignment, aspect ratios within an Adaptation Set, or media durations. MP4Box can help package the content prior to DASHing it. For the most common options we propose some flags during the DASHing phase, such as ```#audio```, ```#video``` or ```:dur=```.

I understand when people blame MP4Box for not checking the media itself for inconsistencies. We would be super happy to provide that but it would require a significant effort from our team. Any idea is warmly welcome.

At the moment the output of MP4Box is correct when the input is correct.

> I disagree that generating empty samples is inappropriate - though I admit it is unreasonably complex for audio/video. For subtitles, it seems a good fit and is, I suspect, also what most users would expect.

3 things:

1) GPAC supports many types of media. Not just audio, video and subtitles but also images, graphics (svg, ...), scenes (bifs, laser, vrml, etc.), nodes, scripts, widgets, ... As a maintainer of such a project, separating the concerns is important. That's why I am conceptually against this change although there have been some exceptions (see 2)).

2) We have re-introduced the behaviour you describe (insertion of void sample) for SRT subtitles, see #362 . Some years ago we introduced the last empty sample trick. Now most players were broken because they expect this last empty sample. The truth is: adding these samples in the end led us to having players which expect a last empty sample i.e. they just ignore the sample duration. What good does that do?

3) GPAC was downloaded hundreds of thousands of times since our implementation of MPEG-DASH. Changing the behaviour may break a lot of existing workflows. We want to make sure we are fixing a real bug and not helping players which are not compliant. Example: last year we had to change the sample duration in the MPD because most players were not able to have separate different segment durations across media - then we got some issues from users who complained of inconsistencies of media duration (and unreliable seeks due to the fact they relied on the timeline only, not on the sidx). That's a lot of work to handle.

> most players encounter serious problems when there are 404s in a DASH presentation.

It inspires me two remarks:
1) Maybe it requires a clarification.
2) Probably there should be some publicly available samples for this use-case with an expected bahaviour based on 1).

What do you think?

**Comment by rbouqueau** [comment]
> However, I note that when I do specify a subtitle duration, it simply duplicates the last segment forever. This seems wrong - the subtitle segments at the end should be empty.

Yes but we can't have a segment with no sample. So we duplicate the last sample and add the  redundancy flag. See http://www.gpac-licensing.com/2016/03/21/guidelines-on-how-to-embed-ttml-in-mp4-and-dash/ for more info. Having an empty sample would make it smaller for sure (but not for TTML though).

**Comment by sandersaares** [comment]
> I don't understand what is "inconsistent". The period duration is the longest duration of any media. What did I miss?

Yes, this was my misunderstanding about current behavior - the source material I used did not have a fixed GOP size, which was throwing off my calculations based on segment numbers and giving me the incorrect impression there was more logic executed than there really was.

Regarding expected player behavior - in general, the expectation I hear from player vendors is that there are no 404s in VOD content. While I tried approaching some about supporting it, they seemed to view this as an edge case not worth bothering about. Indeed, when I brought up the discussion in DASH-IF, the leading opinion was, roughly speaking, that 404s have no place in valid content. CMAF also appears to imply this.

I understand your wish to keep mp4box functionality minimal and constrained. Perhaps let's explore the topic from the viewpoint of how to best use existing mp4box functionality and achieve a good DASH presentation with no 404s.

Assuming subtitles that end early, what would be the correct way to pre-process them and provide them to mp4box to ensure this? How do I get mp4box to emit empty WebVTT/TTML samples after the "normal" duration of the stream? The expected output would be an "empty cue" sample for WebVTT and an empty TTML document sample for TTML.

**Comment by sandersaares** [comment]
I attempted to work around the existing limitations by appending a subtitle sample "near infinity" (very large timestamp) to add padding to the subtitle stream, and manually limiting the subtitle input duration to match audio/video. However, the output still appears inconsistent.

For WebVTT input, the segments for the "overtime" part are still generated, though the manifest is OK (clamped to specified duration) and the segment contents are OK (only an empty cue sample in the padding segments).

For TTML, the padding segments are not generated at all. The manifest is as expected (clamped to specified duration) and the non-padding segment contents are OK.

I am attaching the additional input files I used for testing this.

[padded_subtitles.zip](https://github.com/gpac/gpac/files/279097/padded_subtitles.zip)

**Comment by ulrichghislain** [comment]
Hello i need a help please , i'am a beginner on dashcast 
i have a stream mpeg2ts and i want to have a stream dashcast 
this is command that i'm typing
"DashCast -av udp://@239.x.x.x:1234?fifo_size=3000000 -live -seg-dur 8000 -mpd-refresh 8 -time-shift -1 -min-buffer 10 -npts -conf config_dashcast.conf -ast-offset 16000" then i just have a folder output empty (no chunk stream inside)

i get this error in command ---->"[mpegts @ 0x24561e0] Could not find codec parameters for stream 6 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options"

----->"Assertion desc failed at libswscale/swscale_internal.h:674
thanks in advance

**Comment by rbouqueau** [comment]
For reader' information: we push a MPEG contribution that should be included in an amendment very soon. The next version of MPEG-DASH will include it.

The contribution allows to put a duration at the AS and Rep levels in case the media doesn't have the exact same duration as the Period.

**Comment by joeyparrish** [comment]
Having it in the DASH spec doesn't make it interoperable.  I think it would be greatly preferable for everyone if you were to generate interoperable content.

**Comment by rbouqueau** [comment]
@joeyparrish I don't get the point of your message. How can we improve MP4Box? How can we improve interoperability?

Current state:
 - MP4Box still seems to be compliant with the DASH spec in any case.
 - To help the authoring, we have added a ':dur' option to align the durations.
 - Since some players have issues with unexpected 404s, we pushed an amendment to the DASH standard to add a signalling of different content durations withing Periods.

At this point, my view is that the players don't implement the DASH standard and expect MP4Box to create content that is magically compatible with their limitations: avc1 versus avc3, segment numbers having to be aligned, aspect ratios that are "almost" the same, etc. We probably need more corner case vectors that players' implementors can use.

**Comment by joeyparrish** [comment]
I'm sorry, I didn't mean to come off so rudely.  Please accept my apologies.

What I meant to say is that in my experience, there is much which is permitted in the DASH spec, but interoperability with other components (encoders, packagers, servers, CDNs, players) requires a much more restrictive view of the DASH spec.  We have found it very helpful to follow the DASH IOP rather than the complete spec.

The policy at Shaka Player is that we only work to support non-IOP DASH when doing so is trivial and does not complicate the code.  I believe it would be beneficial to the adoption of DASH and to the health of the DASH ecosystem as a whole if interoperability was a primary goal for projects such as yours and mine.  People expect to take open-source components off the shelf and see them work together, and I think that's a reasonable expectation.

**Comment by sandersaares** [comment]
I support Joey's view that interoperability is key - DASH allows for quite a lot that is not practical in actual real world usage.

Specifically in this case, players tend to have problems playing partial content - if you start playing audio and video, one cannot just stop before the other does. The media pipeline will often stall if you stop feeding one while the other continues.

This problem can be relatively intractable, as players are generally built upon OS/browser media stacks and player vendors often do not have the latitude to even affect how the media stack works. Even if "just don't play audio if the representation ends before the video ends" may be the desired effect from the content author's viewpoint, it is not a reality that can be easily implemented due to A/V sync and decoder assumptions that have a strong basis in the media components.

Therefore, interoperable DASH should not have uneven representations. This has been the common interpretation in DASH-IF when the topic has come up for discussion, though perhaps we could state it even more clearly. I believe the latest CMAF draft does have wording that makes it more explicit for CMAF, though (that representations, or in CMAF terms tracks in the same switching set, must contain segments for the entire duration).

For mp4box to operate in the optimal manner, it should ensure that no uneven representations exist in the DASH it generates (at least if conforming to a profile specified in DASH-IF IOP).

There are multiple hypothetical possibilities for implementation: for audio and video tracks, mp4box might do sample-level cutting to equalize matters, for example. In practice, even segment-level equality might be "close enough" in many situations (though multi-period content is likely more sensitive toward gaps between periods; I have minimal insight into this aspect, though). For text tracks, it is normal for them to "end" early in the source content (TTML or VTT files) but the expected interpretation is that this does not signal "end of the movie" but just "there is no text to be displayed after this". I believe the latter is the specific situation I filed this issue about - that empty text samples are expected to pad out text streams, ensuring no uneven representations and no cutting of content to the last subtitle.

The ability to manually override the duration in mp4box is potentially useful as a workaround (really, mp4box should do it without asking the user to "massage" the data and to perform his own timing analysis) but in my comment https://github.com/gpac/gpac/issues/521#issuecomment-221206009 I noted that this does not actually seem to work all that reliably.

Furthermore, when investigating the issue I ran into a few more issues that complicate matters, that I summarize here:

* https://github.com/gpac/gpac/issues/554 - durations sometimes simply do not add up.
* https://github.com/gpac/gpac/issues/553 - a text adaptation set is sometimes just an empty XML element (despite having valid samples); this was discovered during text timing experiments - I do not know the exact correlation but it showed up in multiple tests
